File: CSC330/poker/C#/poker.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

class Program
{
    static List<Player> ReadHandsFromFile(string filename)
    {
        List<Player> players = new List<Player>();

        try
        {
            using (StreamReader file = new StreamReader(filename))
            {
                string line;
                while ((line = file.ReadLine()) != null)
                {
                    Player player = new Player();
                    string[] cardStrs = line.Split(',');
                    foreach (string cardStr in cardStrs)
                    {
                        string trimmedCardStr = cardStr.Trim();
                        Card card = new Card(trimmedCardStr);
                        player.DealCard(card);
                    }
                    players.Add(player);
                }
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error opening file: {filename}\n{ex.Message}");
            Environment.Exit(1);
        }

        return players;
    }

bool comparePlayers(Player p1, Player p2)
{
    if (p1.HandRank() == p2.HandRank())
    {
        var hand1 = p1.GetHand();
        var hand2 = p2.GetHand();

        for (int i = 0; i < hand1.Count; i++)
        {
            if (hand1[i].GetRankValue() != hand2[i].GetRankValue())
            {
                return hand1[i].GetRankValue() > hand2[i].GetRankValue();
            }
            // If ranks are equal, compare suits in Spades, Hearts, Clubs, Diamonds order
            if (hand1[i].GetSuitValue() != hand2[i].GetSuitValue())
            {
                return hand1[i].GetSuitValue() < hand2[i].GetSuitValue();
            }
        }
        return false; // If all cards are equal
    }
    return p1.HandRank() > p2.HandRank();
}

    static void Main(string[] args)
    {
        List<Player> players;

        if (args.Length == 1)
        {
            players = ReadHandsFromFile(args[0]);
        }
        else
        {
	    Console.WriteLine("*** Random Deck ***");
            Deck deck = new Deck();
            deck.Shuffle();
	    deck.PrintDeck();
	    Console.WriteLine();
            players = new List<Player>();

            for (int i = 0; i < 6; i++)
            {
                Player player = new Player();
                for (int j = 0; j < 5; j++)
                {
                    player.DealCard(deck.Deal());
                }
                players.Add(player);
            }
        }

        foreach (Player player in players)
        {
            player.SortByRank();
        }
	Console.WriteLine("Player Hands");
	for ( int i = 0; i < players.Count; i++) {
		players[i].PrintHand();
		Console.WriteLine();
	}
// Sort players by hand rank, then by card ranks, then by suits
 	players = players.OrderByDescending(p => p.HandRank())
                  .ThenByDescending(p => p.GetHand().Select(c => c.GetRankValue()).FirstOrDefault())
                  .ThenByDescending(p => p.GetHand().Select(c => c.GetSuitValue()).FirstOrDefault())
                  .ToList();
	Console.WriteLine("Player Hands by Rank");
        for (int i = 0; i < players.Count; i++)
        {
            players[i].PrintHand();
            Console.Write("- ");
            switch (players[i].HandRank())
            {
                case 1: Console.Write("High Card"); break;
                case 2: Console.Write("One Pair"); break;
                case 3: Console.Write("Two Pair"); break;
                case 4: Console.Write("Three of a Kind"); break;
                case 5: Console.Write("Straight"); break;
                case 6: Console.Write("Flush"); break;
                case 7: Console.Write("Full House"); break;
                case 8: Console.Write("Four of a Kind"); break;
                case 9: Console.Write("Straight Flush"); break;
            }
            Console.WriteLine();
        }
    }
}

// Comparer class for sorting cards by rank and suit
class CardComparer : IComparer<Card>
{
    public int Compare(Card x, Card y)
    {
        int rankComparison = y.GetRankValue().CompareTo(x.GetRankValue());
        if (rankComparison == 0)
        {
            return x.GetSuitValue().CompareTo(y.GetSuitValue());
        }
        return rankComparison;
    }
}

File: CSC330/poker/Java/Poker.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class Poker {
    public static List<Player> readHandsFromFile(String filename) {
        List<Player> players = new ArrayList<>();

        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                Player player = new Player();
                String[] cards = line.split(",");
                for (String cardStr : cards) {
                    player.dealCard(new Card(cardStr.trim()));
                }
                players.add(player);
            }
        } catch (IOException e) {
            System.err.println("Error reading file: " + e.getMessage());
        }

        return players;
    }

    public static void main(String[] args) {
        List<Player> players;

        if (args.length == 1) {
	    System.out.println("*** Using Test Deck  ***");
	    System.out.println("File - " + args[0]);
            players = readHandsFromFile(args[0]);
        } else {
	    System.out.println("*** Using Random Deck ***");
            Deck deck = new Deck();
            deck.shuffle();
	    deck.printDeck();
            players = new ArrayList<>();

            for (int i = 0; i < 6; i++) {
                Player player = new Player();
                for (int j = 0; j < 5; j++) {
                    player.dealCard(deck.deal());
                }
                players.add(player);
            }
        }
	for (int i = 0; i < 6; i++) {
		players.get(i).printHand();
		System.out.println();
	}
        for (Player player : players) {
            player.sortByRank();
        }

        players.sort(Comparator.comparingInt(Player::handRank).reversed());
	System.out.println("Winning Hands by Type");
        for (int i = 0; i < players.size(); i++) {
            String handType = new String();

            if(players.get(i).handRank() == 1) {
		handType = "High Card";
	    }
            else if(players.get(i).handRank() == 2) {
	    	handType = "One Pair";
	    }
            else if(players.get(i).handRank() == 3) {
                handType = "Two Pair";
            }
	    else if(players.get(i).handRank() == 4) {
                handType = "Three of a Kind";
            }
	    else if(players.get(i).handRank() == 5) {
                handType = "Straight";
            }
	    else if(players.get(i).handRank() == 6) {
                handType = "Flush";
            }
	    else if(players.get(i).handRank() == 7) {
                handType = "Full House";
            }
	    else if(players.get(i).handRank() == 8) {
                handType = "Four of a Kind";
            }
	    else if(players.get(i).handRank() == 9) {
                handType = "Straight Flush";
            }
	    else if(players.get(i).handRank() == 10) {
                handType = "Royal Flush";
            }
            players.get(i).printHand();
	    System.out.print("- " + handType + "\n");
	}
    }
}

